File: src/main/java/com/hazza/LeetCodeByJava/No92ReverseLinkedListII.java
package com.hazza.LeetCodeByJava;

import com.hazza.LeetCodeByJava.datastructure.ListNode;

/**
 * Created with IntelliJ IDEA.
 * Description: https://leetcode.com/problems/reverse-linked-list-ii/description/
 * Author: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 2017-09-29
 * Time: 12:30 PM
 */
public class No92ReverseLinkedListII {

    // assume 1 ≤ m ≤ n ≤ length of list
    public ListNode reverseBetween(ListNode head, int m, int n) {
        if (m == n) return head;
        ListNode start = new ListNode(0);
        start.next = head;
        int count = 0;
        ListNode prev = start;
        while (++count < m) prev = prev.next;
        ListNode cur = prev.next, right = cur.next;
        while (++count <= n) {
            ListNode temp = right.next;
            right.next = cur;
            cur = right;
            right = temp;
        }
        ListNode temp = prev.next;
        prev.next = cur;
        temp.next = right;

        return start.next;
    }
}

File: src/main/python/leetcode_by_python/linked_list/No92_Reverse_Linked_List_II.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019-08-01
#https://leetcode.com/problems/reverse-linked-list-ii/
from src.main.python.leetcode_by_python.data_structure.ListNode import ListNode


class Solution:
    def reverseBetween(self, head: ListNode, m: int, n: int) -> ListNode:
        dummy = ListNode(0)
        dummy.next = head
        prev = dummy

        for _ in range(m - 1):
            prev = prev.next

        cur = prev.next
        right = cur.next
        for _ in range(n - m):
            right.next, cur, right = cur, right, right.next

        prev.next.next, prev.next = right, cur

        return dummy.next



